### PR TITLE
Update pod resource metrics KEP for beta with PRR

### PR DESCRIPTION
**Pull request by smarterclayton**
Pod resource metrics will go to beta in 1.21

### CODE REVIEW

File: keps/sig-instrumentation/1748-pod-resource-metrics/kep.yaml
Comment (dashpole) [inline keps/sig-instrumentation/1748-pod-resource-metrics/kep.yaml:9]: IIRC, they reserve `implemented` to mean GA

**Comment by dashpole** [comment]
This doesn't need a PRR approver since it is already marked implementable, right?

**Comment by ehashman** [comment]
@dashpole when we update the latest milestone to beta, it'll trigger a PRR approver in CI.

File: keps/sig-instrumentation/1748-pod-resource-metrics/kep.yaml
Comment (ehashman) [reply keps/sig-instrumentation/1748-pod-resource-metrics/kep.yaml:9]: Yeah this should stay in `implementable`, we need to add a stage here:

https://github.com/kubernetes/enhancements/blob/6c2e020d4475d4170ffdeafb266a84bb444ac9a6/keps/NNNN-kep-template/kep.yaml#L26-L27

Comment (justaugustus) [reply keps/sig-instrumentation/1748-pod-resource-metrics/kep.yaml:9]: FYI -- I'm working on cleaning up the validation logic in https://github.com/kubernetes/enhancements/pull/2409.

ref: https://kubernetes.slack.com/archives/C1L57L91V/p1612288049074600, https://kubernetes.slack.com/archives/C1L57L91V/p1612385943085000

**Comment by annajung** [comment]
Hi @smarterclayton, 1.21 Enhancements Lead here. 
Please make sure `kep.yaml` reflects the [latest template](https://github.com/kubernetes/enhancements/blob/master/keps/NNNN-kep-template/kep.yaml) which includes `prr-approvers`. 

Also, please make sure to follow the [instruction for getting PRR approved](https://github.com/kubernetes/community/blob/master/sig-architecture/production-readiness.md#submitting-a-kep-for-production-readiness-approval). Specifically, this PR is missing a requirement for a yaml file under https://github.com/kubernetes/enhancements/tree/master/keps/prod-readiness/sig-instrumentation. 

Once you have the requirements, please reach out to get this reviewed by a PRR approval by reaching out in #prod-readiness slack channel asap.

Enhancements freeze is just around the corner: Feb. 9th. Please make sure to get all requirements completed before then. Thank you!

File: keps/sig-instrumentation/1748-pod-resource-metrics/kep.yaml
Comment (smarterclayton) [reply keps/sig-instrumentation/1748-pod-resource-metrics/kep.yaml:9]: I'll change it

**Review by ehashman — APPROVED** [review]
/lgtm
/approve

Needs PRR as well.

**Review by ehashman — COMMENTED** [review]
/lgtm

only diff is @johnbelamaric assigned for PRR

File: keps/sig-instrumentation/1748-pod-resource-metrics/README.md
Comment (johnbelamaric) [inline keps/sig-instrumentation/1748-pod-resource-metrics/README.md:451]: That's not really disabling it. So, if for example this new endpoint turns out to have a high cost when scraped, and can therefore be used as a DOS attack on the scheduler, there is no way to turn off the endpoint?

File: keps/sig-instrumentation/1748-pod-resource-metrics/README.md
Comment (johnbelamaric) [inline keps/sig-instrumentation/1748-pod-resource-metrics/README.md:465]: That's not disabling it, it's just not using it.

File: keps/sig-instrumentation/1748-pod-resource-metrics/README.md
Comment (johnbelamaric) [inline keps/sig-instrumentation/1748-pod-resource-metrics/README.md:488]: how is the metrics endpoint protected, couldn't anyone just scrape it?

File: keps/sig-instrumentation/1748-pod-resource-metrics/README.md
Comment (johnbelamaric) [inline keps/sig-instrumentation/1748-pod-resource-metrics/README.md:518]: Can you list any specific ones?

File: keps/sig-instrumentation/1748-pod-resource-metrics/README.md
Comment (johnbelamaric) [inline keps/sig-instrumentation/1748-pod-resource-metrics/README.md:523]: So no change to any existing scheduler SLOs, presumably?

File: keps/sig-instrumentation/1748-pod-resource-metrics/README.md
Comment (johnbelamaric) [inline keps/sig-instrumentation/1748-pod-resource-metrics/README.md:599]: Is there any way (from metrics or logs) to tell how often the metrics endpoints are being scraped?

Comment (smarterclayton) [reply keps/sig-instrumentation/1748-pod-resource-metrics/README.md:599]: The prometheus default metrics handler should report it (to the scraper), it would show up in a profile, but it does not write a log line (kind of in keeping with our expectations around kube-apiserver).

File: keps/sig-instrumentation/1748-pod-resource-metrics/README.md
Comment (smarterclayton) [reply keps/sig-instrumentation/1748-pod-resource-metrics/README.md:523]: Measuring on a production batch job cluster (prow jobs, o(1k pods)) with this change CPU use on scheduler went from 6m +/- 5m to 7m +/- 5m so basically noise.  Scrape interval for that for is 15s.  Testing with synthetic data this code path has so few allocations that the write time / copy time to the http scraper connection dominates CPU use, which is usually a good sign.

File: keps/sig-instrumentation/1748-pod-resource-metrics/README.md
Comment (smarterclayton) [reply keps/sig-instrumentation/1748-pod-resource-metrics/README.md:488]: It's protected same as the /metrics endpoint which is with delegated auth `--authorization-kubeconfig` which is an RBAC check

File: keps/sig-instrumentation/1748-pod-resource-metrics/README.md
Comment (smarterclayton) [reply keps/sig-instrumentation/1748-pod-resource-metrics/README.md:451]: The cost when scraped is the same as metrics, so the existing behavior for metrics is covered.  We do not allow metrics to be disabled (but if you disable the metrics set pre-GA it's roughly disablement).  You can disable the entire scheduler HTTP server.

I can clarify, but we generally don't make metrics optional and RBAC would be one way of preventing access to metrics, as would be having the scheduler not listen.  Anyone who can DOS the scheduler can do so today using the existing metrics, so we aren't adding a new DOS angle.

File: keps/sig-instrumentation/1748-pod-resource-metrics/README.md
Comment (smarterclayton) [reply keps/sig-instrumentation/1748-pod-resource-metrics/README.md:518]: In this case I would probably look at scheduler up / down from the scraping endpoint, then scheduler cpu.  Is that what you're looking for?

File: keps/sig-instrumentation/1748-pod-resource-metrics/README.md
Comment (johnbelamaric) [reply keps/sig-instrumentation/1748-pod-resource-metrics/README.md:451]: Ok, can you just say that (disable the scheduler http server) and how to do it?

File: keps/sig-instrumentation/1748-pod-resource-metrics/README.md
Comment (johnbelamaric) [reply keps/sig-instrumentation/1748-pod-resource-metrics/README.md:518]: Sure, or you can just say: Existing scheduler SLIs.

File: keps/sig-instrumentation/1748-pod-resource-metrics/README.md
Comment (johnbelamaric) [reply keps/sig-instrumentation/1748-pod-resource-metrics/README.md:523]: Ok, you can just say that: Existing scheduler SLO.

File: keps/sig-instrumentation/1748-pod-resource-metrics/README.md
Comment (johnbelamaric) [reply keps/sig-instrumentation/1748-pod-resource-metrics/README.md:599]: Ok, so if there is a metric from the default metrics handler please list it. This is for operators, so your instructions are more for someone developing it. I think the actions here would be:
- check the metric to see if it's being scraped like crazy
- if so, disable the scraper
- last resort, disable the http server for scheduler metrics, but that has obvious bad consequences for visibility

Comment (smarterclayton) [reply keps/sig-instrumentation/1748-pod-resource-metrics/README.md:599]: Turns out we don't have those metrics on for request count.  Have we tried to impose a general statement about that?  We're kind of inconsistent but i'd expect sig-instrumentation to blanket enforce that (components that expose metrics should also expose the standard prometheus http request handler).  There are some cardinality concerns.  For right now we'd have to add something to catch that, and we don't *generally* do it.  Would you prefer me to provide a simpler path, or can we punt this to a broader "let's have better standard operational metrics for counting requests that don't blow out cardinality"?

Comment (johnbelamaric) [reply keps/sig-instrumentation/1748-pod-resource-metrics/README.md:605]: we can punt it, maybe create an issue for sig-instrumentation?

File: keps/sig-instrumentation/1748-pod-resource-metrics/README.md
Comment (johnbelamaric) [reply keps/sig-instrumentation/1748-pod-resource-metrics/README.md:522]: nit: you could just remove the template text and put the answer

**Review by johnbelamaric — COMMENTED** [review]
/approve

**Pull request merged**
